### PR TITLE
[#10944] Migrate Blobstore API to Google Cloud Storage

### LINKED ISSUE #10944: Migrate Blobstore API to Google Cloud Storage
Similar to #10929 but for this issue it is `com.google.appengine.tools.cloudstorage.*` and `com.google.appengine.api.blobstore.*` which should be changed to using Google Cloud Storage API directly. As before, no additional setup is necessary since Blobstore API uses GCS behind the screen.

### PR DESCRIPTION
**Pull request by wkurniawan07**
Fixes #10944

Compared to #10930 this has a slightly more widespread effect:
- The proxy to GCS is moved to logic API layer to allow mocking, just like how the email sender works. Previously it was deep in the common/utility layer as the Blobstore API is a "taken for granted" API, which is no longer the case.
  - Because of this, some logic needs to be reworked, e.g. the operation to add/remove profile picture is now decoupled from the DB layer and called directly by the API layer.
- Branched implementation: save the file locally in dev server/testing, connect to GCS in staging server.
  - Again, it was also done this way before but hidden behind the screen by the Blobstore API.
  - However, because the implementation is no longer black-boxed, we now know the exact behaviour and can apply more testing, e.g. testing for actual image content instead of just the "blob key".
- Removal of `blobKey` field in `StudentProfile`.
  - We save the profile pictures with the account's google ID as the identifier in GCS. That means we don't need to store any kind of "key" per se in order to retrieve the profile picture; this key was only necessary because of Blobstore API's implementation.
  - The existence of `blobKey` is an indicator of presence/absence of profile picture, but querying GCS directly is possible and not expensive ($0.004 per 10K operations; about the same rate as Datastore querying). In addition we no longer need to update the `StudentProfile` entity whenever profile picture is added/removed, which translates to some cost saving.

### CODE REVIEW

**Comment by madanalogy** [comment]
@jianhandev @Derek-Hardy it might be that this PR will help address some of the test failures you're seeing in the objectify migration (see https://github.com/objectify/objectify/issues/442). Just fyi that it might be worthwhile to merge from master into the feature branch after this gets merged into master. If either of you are up for it you could even try to review this PR :P

File: src/main/java/teammates/logic/core/LocalFileStorageService.java [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+        File file = new File(constructFilePath(fileKey));
+        file.delete();
+    }
+
+    @Override
+    public void create(String fileKey, byte[] contentBytes, String contentType) {
+        try (OutputStream os = Files.newOutputStream(Paths.get(constructFilePath(fileKey)))) {
+            os.write(contentBytes);
+        } catch (IOException e) {
+            // TODO
```
Comment (madanalogy) [inline src/main/java/teammates/logic/core/LocalFileStorageService.java:32]: Maybe just use the logger for this?

File: src/test/java/teammates/test/FileHelper.java [hunk 2]
```diff
@@ -44,7 +45,15 @@ public static void saveFile(String filePath, String content) throws IOException
...
         }
+    }
 
+    /**
+     * Saves the supplied content to the specified file path.
+     */
+    public static void saveFile(String filePath, byte[] content) throws IOException {
+        try (OutputStream os = Files.newOutputStream(Paths.get(filePath))) {
+            os.write(content);
+        }
```
Comment (madanalogy) [inline src/test/java/teammates/test/FileHelper.java:56]: Similarly, possibly log any exceptions that might get thrown here

**Review by madanalogy — COMMENTED** [review]
Sorry for the delay, finally found the time to tackle this. Can't say I completely understand the changes 100% but they look alright to me.

File: src/main/java/teammates/logic/core/LocalFileStorageService.java [hunk 3]
```diff
@@ -1,28 +1,57 @@
...
-        // TODO implement this
-        return null;
+    public boolean doesFileExist(String fileKey) {
+        try {
+            Files.newInputStream(Paths.get(constructFilePath(fileKey)));
+            return true;
+        } catch (IOException e) {
+            return false;
+        }
     }
```
Comment (Derek-Hardy) [inline src/main/java/teammates/logic/core/LocalFileStorageService.java:44]: Do we need to close the `InputStream` in a `finally` block? Similarly for other functions in this file.

File: src/main/java/teammates/logic/core/LocalFileStorageService.java (hunk 1 above)
Comment (wkurniawan07) [reply src/main/java/teammates/logic/core/LocalFileStorageService.java]: I don't see it as that useful, but nothing to lose by doing it

File: src/main/java/teammates/logic/core/LocalFileStorageService.java (hunk 3 above)
Comment (wkurniawan07) [reply src/main/java/teammates/logic/core/LocalFileStorageService.java:43]: The method you're commenting on has changed, but all others are in the form of `try (...) { ... }` which is [try-with-resources](https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html).

**Review by madanalogy — APPROVED** [review]
LGTM!

**Review by Derek-Hardy — APPROVED** [review]
LGTM!

**Pull request merged**